k-distribution/k-tutorial/1_basic/11_casts/README.md

```diff
@@ -0,0 +1,188 @@
+covered in [Lesson 1.4](../04_disambiguation/README.md), K provides one
+additional powerful tool for disambiguation: cast expressions.
+
+K provides three main types of casts: the semantic cast, the strict cast, and
+the projection cast. We will cover each of them, and their similarities and
+differences, in turn.
+
+### Semantic casts
+
+The most basic, and most common, type of cast in K is called the
+**semantic cast**. For every sort `S` declared in a module, K provies the
+following (implicit) production for use in sentences:
+
+```
+  syntax S ::= S ":S"
+```
+
+Note that `S` simply represents the name of the sort. For example, if we 
+defined a sort `Exp`, the actual production for that sort would be:
+
+```
+  syntax Exp ::= Exp ":Exp"
+```
+
+At runtime, this expression will not actually exist; it is merely an annotation
+to the compiler describing the sort of the term inside the cast. It is telling
+the compiler that the term inside the cast must be of sort `Exp`. For example,
+if I had the following grammar:
+
+```k
+module LESSON-11-A
+  imports INT
+
+  syntax Exp ::= Int | Exp "+" Exp
+  syntax Stmt ::= "if" "(" Exp ")" Stmt | "{" "}"
+endmodule
+```
+
+Then we would be able to write `1:Exp`, or `(1 + 2):Exp`, but not `{}:Exp`.
+
+You can also restrict the sort that a variable in a rule will match by casting
+it. For example, consider the following additional module:
+
+```k
+module LESSON-11-B
+  imports LESSON-11-A
+  imports BOOL
+
+  syntax Term ::= Exp | Stmt
+  syntax Bool ::= isExpression(Term) [function]
+
+  rule isExpression(E:Exp) => true
+  rule isExpression(_) => false [owise]
+endmodule
+```
+
+Here we have defined a very simple function that decides whether a term is
+an expression or a statement. It does this by casting the variable inside the
+`isExpression` rule to sort `Exp`. As a result, that variable will only match terms
+of sort `Exp`. Thus, `isExpression(1)` will return true, as will `isExpression(1 + 2)`, but
+`isExpression({})` will return false.
+
+#### Exercise
+
+Verify this fact for yourself by running `isExpression` on the above examples. Then
+write an `isStmt` function, and test that it works as expected.
+
+### Strict casts
+
+On occasion, a semantic cast is not strict enough. It might be that you want
+to, for disambiguation purposes, say **exactly** what sort a term is. For
+example, consider the following definition:
+
+```k
+module LESSON-11-C
+  imports INT
+
+  syntax Exp ::= Int | Exp "+" Exp
+  syntax Exp2 ::= Exp | Exp2 "+" Exp2
+endmodule
+```
+
+This grammar is a little ambiguous and contrived, but it serves to demonstrate
+how a semantic cast might be insufficient to disambiguate a term. If I were 
+to write the term `(I1:Int + I2:Int):Exp2`, the term would be ambiguous,
+because the cast is not sufficiently strict to determine whether you mean
+the derivation `Exp2 -> Exp -> Exp "+" Exp` or the derivation
+`Exp2 -> Exp2 "+" Exp2`.
```

Comment:
Are these what you meant to write? Neither makes sense to me. Perhaps you should just annotate the productions with names `syntax Exp ::= Int | Exp "+" Exp [exp]`, and the say "whether you mean the derivation using the `exp` production or the `exp2` production.

I guess the `_ -> _` notation is confusing to me, because it's not defined anywhere and maybe users won't be familiar with it from other contexts. Whereas we've already explained putting user-defined attributes on productions and using them to identify them.